File: src/creational_patterns/abstract_factory/java/abstract_factory.java
interface Shape {
   void draw();
}

class Circle implements Shape {
   @Override
   public void draw() {
      System.out.println("Drawing Circle");
   }
}

class Rectangle implements Shape {
   @Override
   public void draw() {
      System.out.println("Drawing Rectangle");
   }
}

class Square implements Shape {
   @Override
   public void draw() {
      System.out.println("Drawing Square");
   }
}

interface AbstractFactory {
   Shape getShape(String shapeType);
}

class ShapeFactory implements AbstractFactory {
   @Override
   public Shape getShape(String shapeType) {
      if (shapeType == null) {
         return null;
      }
      if (shapeType.equalsIgnoreCase("CIRCLE")) {
         return new Circle();
      } else if (shapeType.equalsIgnoreCase("RECTANGLE")) {
         return new Rectangle();
      } else if (shapeType.equalsIgnoreCase("SQUARE")) {
         return new Square();
      }
      return null;
   }
}

class FactoryProducer {
   public static AbstractFactory getFactory(String choice) {
      if (choice.equalsIgnoreCase("SHAPE")) {
         return new ShapeFactory();
      }
      return null;
   }
}

public class AbstractFactoryDemo {
   public static void main(String[] args) {
      AbstractFactory shapeFactory = FactoryProducer.getFactory("SHAPE");
      Shape shape1 = shapeFactory.getShape("CIRCLE");
      shape1.draw();
      Shape shape2 = shapeFactory.getShape("RECTANGLE");
      shape2.draw();
      Shape shape3 = shapeFactory.getShape("SQUARE");
      shape3.draw();
   }
}

File: src/creational_patterns/abstract_factory/php/abstract_factory.php
interface Shape {
    public function draw();
}

class Circle implements Shape {
    public function draw() {
        echo "Drawing a Circle\n";
    }
}

class Square implements Shape {
    public function draw() {
        echo "Drawing a Square\n";
    }
}

interface Color {
    public function fill();
}

class Red implements Color {
    public function fill() {
        echo "Filling with Red\n";
    }
}

class Green implements Color {
    public function fill() {
        echo "Filling with Green\n";
    }
}

abstract class AbstractFactory {
    abstract function getShape($shapeType);
    abstract function getColor($colorType);
}

class ShapeFactory extends AbstractFactory {
    public function getShape($shapeType) {
        if ($shapeType == "CIRCLE") {
            return new Circle();
        } else if ($shapeType == "SQUARE") {
            return new Square();
        }

        return null;
    }

    public function getColor($colorType) {
        return null;
    }
}

class ColorFactory extends AbstractFactory {
    public function getShape($shapeType) {
        return null;
    }

    public function getColor($colorType) {
        if ($colorType == "RED") {
            return new Red();
        } else if ($colorType == "GREEN") {
            return new Green();
        }

        return null;
    }
}

class FactoryProducer {
    public static function getFactory($choice) {
        if ($choice == "SHAPE") {
            return new ShapeFactory();
        } else if ($choice == "COLOR") {
            return new ColorFactory();
        }

        return null;
    }
}

$shapeFactory = FactoryProducer::getFactory("SHAPE");
$shape1 = $shapeFactory->getShape("CIRCLE");
$shape1->draw();

$colorFactory = FactoryProducer::getFactory("COLOR");
$color1 = $colorFactory->getColor("RED");
$color1->fill();
